Sources/RoutingKit/PrefixGroupable.swift

```diff
@@ -0,0 +1,36 @@
+public protocol PrefixGroupable: Router {
```

Comment:
I don't think we need this protocol. We could just add an extension to `Router` that returns the `PrefixedRouter`.

```swift
extension Router {
    func grouped(_ prefix: [PathComponent]) -> PrefixedRouter<Self> {
        return .init(...)
    }
}
```

Then we only need `PrefixedRouter` to have one generic argument, the router type it is prefixing. 

Even then though I'm not a huge fan of this approach. I like that in Vapor we can assume `HTTPRoutes` / `HTTPRouteBuilder` and not need any generic arguments. This lets us keep the route grouping types internal, i.e.:

```swift
extension Router {
    func grouped(_ prefix: [PathComponent]) -> Router {
        return _PrivatePrefixingRouter(...)
    }
}
```

Reply:
Yeah that's a good point. I'll shoot to ensure that hidden types stay hidden. My goal is no public or functional changes to vapor/vapor so that this PR can be backported to Vapor 3 if it's decided that that would be a good idea.

Reply:
I had to make some changes to `Router` in order to get this to work. Lemme know if you think it's worth it and worth keeping.

Reply:
It seems like it won't be possible to keep the `PrefixGroupedRouter ` type internal here. Since `Router` has an associated type, you can't just return `-> Router`, you must do:

```swift
func grouped(_ prefix: [PathComponent]) -> PrefixGroupedRouter<BaseRouterType>
```

That's a deal breaker for me. Continuing to do this in Vapor where we have a protocol without associated types to work with gives us a lot more freedom.

Reply:
Might I ask why the associated type was added to `Router` in the first place? We didn't need it before.

Reply:
The way I have it set up, and the way it's set up in vapor/vapor, creating a route group returns an object with a reference to the base router. All registrations are passed through to that base router. This is necessary because of the way route registration is set up.

```swift
func routes(router: inout Router) { // routes need to be registered by mutating `router`

}
```

There may be a way to get around this requirement if we find some other way to do registration.

We can also get rid of the new `associatedtype` if we assume the base router will always be a `TrieRouter`.

Reply:
@tanner0101 I'll see if I can find a way around that.

Is it necessary that `PrefixGroupedRouter<>` isn't exposed by vapor/routing, or are we good as long as it's not exposed to the users of vapor/vapor?

Reply:
@twof anything with `public` in front of it becomes maintenance burden, whether or not it's in the `vapor/vapor` repo.

Reply:
@tanner0101 I've updated it so that the grouping mechanism is built into `TrieRouter`. This means the grouping function can return a `TrieRouter` which was already publicly exposed.